lib/header.c

```diff
@@ -467,15 +468,14 @@ static int dataLength(rpm_tagtype_t type, rpm_constdata_t p, rpm_count_t count,
 	break;
 
     default:
-	if (typeSizes[type] == -1)
-	    return -1;
-	length = typeSizes[(type & 0xf)] * count;
-	if (length < 0 || (se && (s + length) > se))
+	/* this won’t overflow, since ‘long_count’ is 64-bit */
+	length = typeSizes[type] * long_count;
+	if (length <= 0 || (se && length > se - s))
```

Comment:
I don't see a need for an extra variable. You could just assign to length and multiply, or make eg typeSizes int64 type (and/or count itself for that matter)

Note that this ends up removing an old safety feature in the code, ie that `(type & 0xf)` ensures that we'll never access typeSizes out of bounds. Only there was that redundant `== -1` check before it, making it moot :facepalm:

Reply:
I went with making `typeSizes` 64-bit, which will help prevent similar problems elsewhere.